replace signalfx plugin with new one that sends events and produces m…

mdubbyap opened the pull request:
…etadata

mpetazzoni commented:
Good stuff. Consider looking at the `platform` and `netifaces` Python modules though, they'll help clean the code a lot.

wt commented:
I looked over this. It looks good. I agree with max on the platforms module. I am slightly more iffy about the netifaces module since it's not part of the stdlib.

Also, why didn't you use the check_output function from the subprocess module. Are you trying to retain python2.6 compatiblity?

mdubbyap commented:
Platform module throws exceptions while being run within the Python collectd plugin. I'll check the netifaces module. Ack for 2.6 compatibility on using popen vs subprocess. I should add comments on these to avoid confusion.

wt commented:
You can probably work around the exceptions by parsing /proc/version or /proc/version_signature.

mdubbyap commented:
I've added comments that explain the decisions that were commented on.  I don't really like the idea of including netifaces as i would have to package it up and include it here and we're trying not to add any prep steps like using pip.  If i run into compatibility issues with the code i have i will likely have to, but this is fine for now i think. It's a starting point.

I like the idea of parsing /proc/version so long as it exists on all the platforms, but can we merge this so the interns can start packaging this for doghunt?  Thanks.

I've also added a changelog, and a version that we send in with the meta-data log.  I also emit a single datapoint as proof of life.

mpetazzoni commented:
I don't think it's an issue to install other packages via pip. Since we're already in Python, we assume the user has a functioning Python environment, and it's pretty common to ask for `pip install -r requirements.txt`.

We could actually embrace this approach even more and make it so that you don't even have to configure a `ModulePath` in the collectd config, by making this plugin a Python module that is itself pip-installable (and thus would bring in its dependencies automatically).

mpetazzoni commented:
Finally, make sure you `flake8` your code (PEP-8) and write proper Python docstrings (PEP-257).

wt commented on src/signalfx_metadata.py:
```diff
@@ -0,0 +1,255 @@
+    namestr = names.tostring()
+    return [(namestr[i:i + 16].split('\0', 1)[0],
+             socket.inet_ntoa(namestr[i + 20:i + 24]))
+            for i in range(0, outbytes, struct_size)]
```
Will this work for getting more than the first address on an interface?

wt commented on src/signalfx_metadata.py:
```diff
@@ -0,0 +1,255 @@
+        if ipaddress == "127.0.0.1":
+            continue
+        host_info["ipaddress_" + interface] = ipaddress
+        host_info["fqdn_" + interface] = socket.gethostbyaddr(ipaddress)[0]
```
The gethostbyaddr call is obsolete as far a syscalls go.

http://linux.die.net/man/3/gethostbyname

Should you use getaddrinfo instead?

wt commented on src/signalfx_metadata.py:
```diff
@@ -0,0 +1,255 @@
+
+
+def get_cpu_info(host_info={}):
+    info_raw = popen(["cat", "/proc/cpuinfo"])
```
Just open and read the /proc/cpuinfo file instead of catting it.

wt commented on src/signalfx_metadata.py:
```diff
@@ -0,0 +1,255 @@
+    get_aws("ami-id", host_info)
+    get_aws("mac", host_info)
+    get_aws("reservation-id", host_info)
+    get_aws("profile", host_info)
```
You should add the aws region and zone to this.

wt replied:
Oh, and rather than getting a bunch of different requests. Look into using http://169.254.169.254/latest/dynamic/instance-identity/document. It's a json file that includes a bunch of identifying info.

Also, and the following is opinion, I am not sure how interesting "mac" or "reservation-id" or "profile" is in the general case. Reservation-ids are the least useful in that they have nothing to do with RIs. They can be used to identify machines that were started at the same time, but I have never needed that info. "Mac" seems not worth the extra request. "Profile" feels the same way to me.

wt commented on src/signalfx_metadata.py:
```diff
@@ -0,0 +1,255 @@
+
+def get_aws(thing, host_info={}):
+    try:
+        r = requests.get("http://169.254.169.254/latest/meta-data/%s" % thing, timeout=0.1)
```
Requests is not in the python stdlib. Make sure you document the need for a pip install.

mpetazzoni replied:
If you require `requests`, then you might as well require `netifaces` :)

mdubbyap commented:
i've removed any non stdlib package i believe and incorporated all your feedback i believe (except for using non stdlib packages <;) and have included the collectd version.  i've also made it so on start the datapoint is emitted first, then the events on further passes, this fixes a race condition that existed in creating the host dimensions.
